### DISCUSSION TITLE
LateInitializationError with the new NotifierProvider

### DISCUSSION THREAD

**Discussion by anmol-me** [rrousselGit/riverpod#2464] [Q&A] [answered]
**Provider**
```
final homeFeedProvider = NotifierProvider<HomeFeedNotifier, List<News>>(() {
  return HomeFeedNotifier();
});
```
**Class**
```
class HomeFeedNotifier extends Notifier<List<News>> {
  late final UserPreferences userPrefs;

  @override
  List<News> build() {
    userPrefs = ref.watch(userPrefsProvider);
    return [];
  }

Future<void> fetchEntries(BuildContext context) async {
// Data is returned from the internet and put into state list.
}
```

This fetches data from the internet on the first run but when I use the provider again to refresh the method, it gives the following error-
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: LateInitializationError: Field 'userPrefs' has already been initialized.

**Provider/Method used again**
```
    ref.refresh(homeFeedProvider.notifier).fetchEntries(context).then(
          (_) => isLoadingHomePageController.update((state) => false),
        );
```

**Comment by AhmedLSayed9** [comment]
That's expected.
When you refresh a NotifierProvider/AsyncNotifierProvider, its Notifier/AsyncNotifier class will not be recreated (class HomeFeedNotifier), only build will be re-executed.

You can use the dependency using a local variable inside `build`, no need to store it in a variable inside your class:
```dart 
List<News> build() {
    final userPrefs = ref.watch(userPrefsProvider);
    return [];
  }
```
Also, you access it the same way at your side-effect methods by using `ref.read(userPrefsProvider)`.

**Comment by anmol-me** [comment]
The whole point of putting dependency in the class variable is to make it available throughout the class so that we don't have to repeat its initialization. That's what we use to do with StateNotifierProvider. Is this no longer supported?

I found one way of making it work by making dependency in the class variable nullable. Is this the right way to do it?

```
class HomeFeedNotifier extends Notifier<List<News>> {
  UserPreferences? userPrefs;

  @override
  List<News> build() {
    userPrefs = ref.watch(userPrefsProvider);
    }
}
```

**Reply by moxiaov587** [reply, marked as answer]
You can also do that

```dart
class HomeFeedNotifier extends Notifier<List<News>> {
  late UserPreferences userPrefs;

  @override
  List<News> build() {
    userPrefs = ref.watch(userPrefsProvider);
    }
}
```

As @AhmedLSayed9  said
> When you refresh a NotifierProvider/AsyncNotifierProvider, its Notifier/AsyncNotifier class will not be recreated (class HomeFeedNotifier), only build will be re-executed.

The reason for your error is that the final variable is assigned again when you rebuild.
